### PR TITLE
Micro Macro Dictionaries

### PR DESCRIPTION
**Pull request by ActuallyTaylor**
# Overview
This PR adds the initial implementation of a CAN object dictionary abstraction layer.

This layer removes a lot of the difficulty that comes from working with the CANOpen layer of EVT-Core. Specifically it handles:

1. Adds Macros to handle the creation of object dictionaires.
2. Abstracts away the CAN Initialization into 3 discrete functions. This handles creating connections to the CAN Network in the main file.
3. Creates a single function to handle the processing of CAN Nodes.

This PR still needs to be tested but will be tested Thursday.

# Discussion
## Core import into Macro Files
```
// CANOpenMacros.hpp, line 19
#include "co_core.h"
```
Something I want to discuss is importing `co_core.h` into the `CANOpenMacros.hpp` file. This is a macros file so I don't like importing the entirety of CANOpen into the file. Something I think could fix this is importing each type we need individually. I'm not sure of this is the best route though.

## Overall documentation
I want to figure out the best way to document the changes that have happened in this pull request. It changes everything about how we use CAN and every repository will need to change. My current ideas are heavy in code documentation and changes to the ramp-up project. But this does not address the large chunk of the team that has passed through ramp-up and would need to re-learn. I think a presentation on the changes may work, giving the entire team a change to talk about the change.

### CODE REVIEW

**Comment by ActuallyTaylor** [comment]
Its almost ready

**Comment by ActuallyTaylor** [comment]
Added docs

**Comment by mjh9585** [comment]
Instead of including `co_core.h` into the `CANOpenMacros.hpp` you could import `co_obj.h` to limit the scope of the imports. This is the file that has all the constants and structs you use.

**Comment by ActuallyTaylor** [comment]
> Instead of including `co_core.h` into the `CANOpenMacros.hpp` you could import `co_obj.h` to limit the scope of the imports. This is the file that has all the constants and structs you use.

I would like to do that, but co_obj.h only exposes some of the objects. You need to manually import a ton of imports:
```c++
#include "co_integer8.h"
#include "co_integer32.h"
#include "co_emcy_id.h"
#include "co_hb_prod.h"
// ... even more below this
```

File: include/EVT/io/CANOpenMacros.hpp [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+#define PDO_MAPPING_UNSIGNED32 32
+#define PDO_MAPPING_UNSIGNED64 64
+
+#define TRANSMIT_PDO_INHIBIT_TIME_DISABLE 0x00
+
+// TODO: I don't like importing co_core into a macros file. We should discuss importing all the types we need individually.
+#include "co_core.h"
+
+/*!
+ * @brief MANDATORY IDENTIFICATION OBJECTS
```
[the remark this answers is not in the document]
Comment (ActuallyTaylor) [reply include/EVT/io/CANOpenMacros.hpp:20]: I believe this is the standard for Doxygen. I continued the format of documentation from the canopen-stack repository. I can swap it to what we use across the rest of the repository... but after a quick look, it seems like we do this in a good chunk of our functions with @param and @return. We just omit the `!` and `@brief`

File: include/EVT/io/CANOpenMacros.hpp [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+/*!
+ * @brief RPDO SETTINGS OBJECT
+ *
+ * This macro creates an RPDO settings object. This macro itself is abstract,
+ * allowing it to be used with any RPDO number supported by CANOpen. To make
+ * interaction with the entire object dictionary easier you just specify a number
+ * starting at 0. This macro also takes in information about the TPDO that is this
+ * RPDO receives data from. This is the TPDO number and the NODE ID.
+ *
+ * @param RPDO_NUMBER (integer) the RPDO number this settings object is for.
```
[the remark this answers is not in the document]
Comment (ActuallyTaylor) [reply include/EVT/io/CANOpenMacros.hpp:148]: I have looked over the [doxygen manual](https://www.doxygen.nl/manual/) and I can't find anything that says there's an alternative way to document macros. So I believe this is the proper way.

File: include/EVT/io/CANOpenMacros.hpp [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+        /* TPDO #N Settings Object */                                                             \
+        .Key = CO_KEY(0x1800 + TPDO_NUMBER, 0x00, CO_OBJ_D___R_),                                 \
+        .Type = CO_TUNSIGNED8,                                                                    \
+        .Data = (CO_DATA) 0x05,                                                                   \
+    },                                                                                            \
+        {                                                                                         \
+            /* COB-ID used by TPDO  180h+TPDO Node-ID*/                                           \
+            .Key = CO_KEY(0x1800 + TPDO_NUMBER, 0x01, CO_OBJ_DN__R_),                             \
+            .Type = CO_TPDO_ID,                                                                   \
+            .Data = (CO_DATA) CO_COBID_TPDO_DEFAULT(TPDO_NUMBER),                                 \
```
[the remark this answers is not in the document]
Comment (ActuallyTaylor) [reply include/EVT/io/CANOpenMacros.hpp:237]: 1. The NODE ID is included automatically now by the `CO_OBJ_DN__R_` specifically the `N` in this object which adds the NODE_ID deeper in the can library

File: include/EVT/io/CANopen.hpp [hunk 4]
```diff
@@ -22,39 +23,79 @@
...
+ * @param canStackDriver references to the drivers that will be used in initialization
+ * @param sdoBuffer the sdoBuffer
+ * @param appTmrMem the appTmrMem
+ */
+void initializeCANopenNode(CO_NODE* canNode, CANDevice* canDevice, CO_IF_DRV* canStackDriver, uint8_t sdoBuffer[CO_SSDO_N * CO_SDO_BUF_BYTE], CO_TMR_MEM appTmrMem[16]);
+
+/*!
+ * @brief Process events that happened on the given canNode
+ *
+ * Takes in a reference to a CO_NODE and processes events
```
[the remark this answers is not in the document]
Comment (ActuallyTaylor) [reply include/EVT/io/CANopen.hpp:94]: I thought it was just easier to read, but after reading over the `@brief` I think it is appropriate to remove the brief from this function.

File: samples/canopen/canopen_rpdo/main.cpp [hunk 5]
```diff
@@ -63,6 +59,19 @@ int main() {
     // Initialize system
     EVT::core::platform::init();
 
+    // Initialize the timer
+    DEV::Timer& timer = DEV::getTimer<DEV::MCUTimer::Timer2>(100);
+    timer.stopTimer();
```
[the remark this answers is not in the document]
Comment (ActuallyTaylor) [reply samples/canopen/canopen_rpdo/main.cpp:64]: I actually don't know, I believe this is a mistake

File: include/EVT/io/CANOpenMacros.hpp [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+ *
+ */
+#define MANDATORY_IDENTIFICATION_ENTRIES_1000_1014      \
+    {                                                   \
+        /* Device Type */                               \
+        .Key = CO_KEY(0x1000, 0x00, CO_OBJ_____R_),     \
+        .Type = CO_TUNSIGNED32,                         \
+        .Data = (CO_DATA) 0x00,                         \
+    },                                                  \
+        {                                               \
```
[the remark this answers is not in the document]
Comment (mjmagee991) [reply include/EVT/io/CANOpenMacros.hpp:34]: This could still use some work

**Review by mjmagee991 — APPROVED** [review]
One last thing to look at, but great job!

**Review by tmb7852 — APPROVED** [review]

File: include/EVT/io/CANOpenMacros.hpp [hunk 7]
```diff
@@ -0,0 +1,313 @@
...
+#define PDO_MAPPING_UNSIGNED32 32
+#define PDO_MAPPING_UNSIGNED64 64
+
+#define TRANSMIT_PDO_INHIBIT_TIME_DISABLE 0x00
+
+// TODO: I don't like importing co_core into a macros file. We should discuss importing all the types we need individually.
+#include "co_core.h"
+
+/**
+ * This macro helps to create a section of the CANOpen Object dictionary
```
[the remark this answers is not in the document]
Comment (ActuallyTaylor) [reply include/EVT/io/CANOpenMacros.hpp:23]: I think this is a good idea, but what do we think the format should be?

**Pull request merged**